File: TarsEngine/Consciousness/Intelligence/Decision/IntuitiveDecisionMaking.cs
using Microsoft.Extensions.Logging;
using TarsEngine.Consciousness.Intelligence.Pattern;
using TarsEngine.Consciousness.Intelligence.Heuristic;
using TarsEngine.Consciousness.Intelligence.Gut;
using GutDecisionResult = TarsEngine.Consciousness.Intelligence.Gut.DecisionResult;

namespace TarsEngine.Consciousness.Intelligence.Decision;

/// <summary>
/// Implements intuitive decision making capabilities
/// </summary>
public class IntuitiveDecisionMaking
{
    private readonly ILogger<IntuitiveDecisionMaking> _logger;
    private readonly ImplicitPatternRecognition _patternRecognition;
    private readonly HeuristicReasoning _heuristicReasoning;
    private readonly GutFeelingSimulation _gutFeeling;
    private readonly System.Random _random = new();
    private double _intuitiveDecisionLevel = 0.5; // Starting with moderate intuitive decision making
    private readonly List<DecisionRecord> _decisionHistory = new();

    /// <summary>
    /// Gets the intuitive decision level (0.0 to 1.0)
    /// </summary>
    public double IntuitiveDecisionLevel => _intuitiveDecisionLevel;

    /// <summary>
    /// Initializes a new instance of the <see cref="IntuitiveDecisionMaking"/> class
    /// </summary>
    /// <param name="logger">The logger</param>
    /// <param name="patternRecognition">The pattern recognition service</param>
    /// <param name="heuristicReasoning">The heuristic reasoning service</param>
    /// <param name="gutFeeling">The gut feeling simulation service</param>
    public IntuitiveDecisionMaking(
        ILogger<IntuitiveDecisionMaking> logger,
        ImplicitPatternRecognition patternRecognition,
        HeuristicReasoning heuristicReasoning,
        GutFeelingSimulation gutFeeling)
    {
        _logger = logger;
        _patternRecognition = patternRecognition;
        _heuristicReasoning = heuristicReasoning;
        _gutFeeling = gutFeeling;
    }

    /// <summary>
    /// Updates the intuitive decision level
    /// </summary>
    /// <returns>True if the update was successful, false otherwise</returns>
    public bool Update()
    {
        try
        {
            // Gradually increase intuitive decision level over time (very slowly)
            if (_intuitiveDecisionLevel < 0.95)
            {
                _intuitiveDecisionLevel += 0.0001 * _random.NextDouble();
                _intuitiveDecisionLevel = Math.Min(_intuitiveDecisionLevel, 1.0);
            }

            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating intuitive decision making");
            return false;
        }
    }

    /// <summary>
    /// Makes an intuitive decision
    /// </summary>
    /// <param name="decision">The decision description</param>
    /// <param name="options">The options</param>
    /// <param name="domain">The domain (optional)</param>
    /// <returns>The intuitive decision</returns>
    public async Task<IntuitiveDecision> MakeIntuitiveDecisionAsync(string decision, List<string> options, string? domain = null)
    {
        try
        {
            _logger.LogInformation("Making intuitive decision: {Decision}", decision);

            if (options.Count == 0)
            {
                throw new ArgumentException("No options provided for decision");
            }

            // Choose intuition type based on decision characteristics
            var intuitionType = ChooseIntuitionType(decision, domain);

            _logger.LogDebug("Chosen intuition type: {IntuitionType}", intuitionType);

            // Make decision based on intuition type
            object result;

            switch (intuitionType)
            {
                case IntuitionType.PatternRecognition:
                    result = MakePatternBasedDecision(decision, options, domain);
                    break;

                case IntuitionType.HeuristicReasoning:
                    result = _heuristicReasoning.MakeHeuristicDecision(decision, options, domain);
                    break;

                case IntuitionType.GutFeeling:
                    result = _gutFeeling.MakeGutDecision(decision, options);
                    break;

                default:
                    // Default to heuristic reasoning
                    result = _heuristicReasoning.MakeHeuristicDecision(decision, options, domain);
                    break;
            }

            // Create intuitive decision
            var intuitiveDecision = new IntuitiveDecision
            {
                Decision = decision,
                SelectedOption = GetSelectedOption(result),
                Options = options,
                Confidence = GetConfidence(result) * _intuitiveDecisionLevel,
                IntuitionType = intuitionType,
                Timestamp = DateTime.UtcNow,
                Explanation = GenerateDecisionExplanation(result, intuitionType)
            };

            // Record decision
            RecordDecision(intuitiveDecision);

            _logger.LogInformation("Made intuitive decision: {SelectedOption} for {Decision} (Confidence: {Confidence:F2})",
                intuitiveDecision.SelectedOption, intuitiveDecision.Decision, intuitiveDecision.Confidence);

            return intuitiveDecision;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error making intuitive decision");

            // Return basic decision
            return new IntuitiveDecision
            {
                Decision = decision,
                SelectedOption = options.FirstOrDefault() ?? string.Empty,
                Options = options,
                Confidence = 0.3,
                IntuitionType = IntuitionType.GutFeeling,
                Timestamp = DateTime.UtcNow,
                Explanation = "Decision made with low confidence due to an error in the decision-making process"
            };
        }
    }

    /// <summary>
    /// Chooses an intuition type based on decision characteristics
    /// </summary>
    /// <param name="decision">The decision description</param>
    /// <param name="domain">The domain</param>
    /// <returns>The chosen intuition type</returns>
    private IntuitionType ChooseIntuitionType(string decision, string? domain)
    {
        // Check for pattern recognition keywords
        if (decision.Contains("pattern", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("similar", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("recognize", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("familiar", StringComparison.OrdinalIgnoreCase))
        {
            return IntuitionType.PatternRecognition;
        }

        // Check for heuristic reasoning keywords
        if (decision.Contains("principle", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("rule", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("guideline", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("best practice", StringComparison.OrdinalIgnoreCase))
        {
            return IntuitionType.HeuristicReasoning;
        }

        // Check for gut feeling keywords
        if (decision.Contains("feel", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("emotion", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("sense", StringComparison.OrdinalIgnoreCase) ||
            decision.Contains("gut", StringComparison.OrdinalIgnoreCase))
        {
            return IntuitionType.GutFeeling;
        }

        // Check if domain suggests a particular intuition type
        if (!string.IsNullOrEmpty(domain))
        {
            if (domain.Contains("design", StringComparison.OrdinalIgnoreCase) ||
                domain.Contains("architecture", StringComparison.OrdinalIgnoreCase))
            {
                return IntuitionType.PatternRecognition;
            }

            if (domain.Contains("development", StringComparison.OrdinalIgnoreCase) ||
                domain.Contains("coding", StringComparison.OrdinalIgnoreCase) ||
                domain.Contains("programming", StringComparison.OrdinalIgnoreCase))
            {
                return IntuitionType.HeuristicReasoning;
            }

            if (domain.Contains("user", StringComparison.OrdinalIgnoreCase) ||
                domain.Contains("experience", StringComparison.OrdinalIgnoreCase) ||
                domain.Contains("interface", StringComparison.OrdinalIgnoreCase))
            {
                return IntuitionType.GutFeeling;
            }
        }

        // Choose randomly based on current levels
        double patternProb = _patternRecognition.PatternRecognitionLevel * 0.4;
        double heuristicProb = _heuristicReasoning.HeuristicReasoningLevel * 0.4;
        double gutProb = _gutFeeling.GutFeelingLevel * 0.2;

        // Normalize probabilities
        double total = patternProb + heuristicProb + gutProb;
        patternProb /= total;
        heuristicProb /= total;
        gutProb /= total;

        // Choose type based on probabilities
        double rand = _random.NextDouble();

        if (rand < patternProb)
        {
            return IntuitionType.PatternRecognition;
        }
        else if (rand < patternProb + heuristicProb)
        {
            return IntuitionType.HeuristicReasoning;
        }
        else
        {
            return IntuitionType.GutFeeling;
        }
    }

    /// <summary>
    /// Makes a pattern-based decision
    /// </summary>
    /// <param name="decision">The decision description</param>
    /// <param name="options">The options</param>
    /// <param name="domain">The domain</param>
    /// <returns>The decision result</returns>
    private GutDecisionResult MakePatternBasedDecision(string decision, List<string> options, string? domain)
    {
        // Score each option
        var optionScores = new Dictionary<string, double>();

        foreach (var option in options)
        {
            // Recognize patterns in the option
            var patterns = _patternRecognition.RecognizePatterns(option, domain);

            // Calculate score based on pattern matches
            double score = patterns.Count > 0
                ? patterns.Average(p => p.Confidence)
                : 0.3;

            // Add some randomness based on pattern recognition level
            score += (0.2 * (_random.NextDouble() - 0.5)) * _patternRecognition.PatternRecognitionLevel;

            // Ensure score is within bounds
            score = Math.Max(0.1, Math.Min(0.9, score));

            optionScores[option] = score;
        }

        // Choose the option with the highest score
        var selectedOption = optionScores.OrderByDescending(kvp => kvp.Value).First().Key;
        double confidence = optionScores[selectedOption] * _patternRecognition.PatternRecognitionLevel;

        // Create decision result
        var result = new Gut.DecisionResult
        {
            Decision = decision,
            SelectedOption = selectedOption,
            Options = options,
            OptionScores = optionScores,
            Confidence = confidence,
            ReasoningType = "PatternRecognition",
            Timestamp = DateTime.UtcNow
        };

        return result;
    }

    /// <summary>
    /// Generates a decision explanation
    /// </summary>
    /// <param name="result">The decision result</param>
    /// <param name="intuitionType">The intuition type</param>
    /// <returns>The explanation</returns>
    private string GetSelectedOption(object result)
    {
        if (result is Gut.DecisionResult gutResult)
        {
            return gutResult.SelectedOption;
        }
        else if (result is Heuristic.DecisionResult heuristicResult)
        {
            return heuristicResult.SelectedOption;
        }
        return "Unknown";
    }

    private double GetConfidence(object result)
    {
        if (result is Gut.DecisionResult gutResult)
        {
            return gutResult.Confidence;
        }
        else if (result is Heuristic.DecisionResult heuristicResult)
        {
            return heuristicResult.Confidence;
        }
        return 0.5; // Default confidence
    }

    private string GenerateDecisionExplanation(object result, IntuitionType intuitionType)
    {
        switch (intuitionType)
        {
            case IntuitionType.PatternRecognition:
                var patternResult = result as Gut.DecisionResult;
                return $"This decision is based on recognizing patterns in the options. " +
                       $"The selected option '{patternResult?.SelectedOption}' matched familiar patterns with a confidence of {patternResult?.Confidence:F2}.";

            case IntuitionType.HeuristicReasoning:
                var heuristicResult = result as Heuristic.DecisionResult;
                string ruleText = "";
                if (heuristicResult != null && heuristicResult.AppliedRules != null && heuristicResult.AppliedRules.Count > 0)
                {
                    ruleText = $" Key principles applied: {string.Join(", ", heuristicResult.AppliedRules)}.";
                }

                return $"This decision is based on applying heuristic reasoning principles to the options. " +
                       $"The selected option '{GetSelectedOption(result)}' aligned with these principles with a confidence of {GetConfidence(result):F2}.{ruleText}";

            case IntuitionType.GutFeeling:
                var gutResult = result as Gut.DecisionResult;
                return $"This decision is based on a gut feeling about the options. " +
                       $"The selected option '{gutResult?.SelectedOption}' felt right with a confidence of {gutResult?.Confidence:F2}.";

            default:
                return $"This decision was made intuitively with a confidence of {GetConfidence(result):F2}.";
        }
    }

    /// <summary>
    /// Records a decision
    /// </summary>
    /// <param name="decision">The decision</param>
    private void RecordDecision(IntuitiveDecision decision)
    {
        var record = new DecisionRecord
        {
            Decision = decision.Decision,
            SelectedOption = decision.SelectedOption,
            Options = decision.Options,
            Confidence = decision.Confidence,
            IntuitionType = decision.IntuitionType,
            Timestamp = decision.Timestamp,
            Outcome = null // Outcome not known yet
        };

        _decisionHistory.Add(record);
    }

    /// <summary>
    /// Records a decision outcome
    /// </summary>
    /// <param name="decision">The decision</param>
    /// <param name="outcome">The outcome</param>
    /// <param name="success">Whether the decision was successful</param>
    public void RecordDecisionOutcome(string decision, string outcome, bool success)
    {
        // Find the most recent matching decision
        var record = _decisionHistory
            .Where(d => d.Decision == decision && d.Outcome == null)
            .OrderByDescending(d => d.Timestamp)
            .FirstOrDefault();

        if (record != null)
        {
            record.Outcome = outcome;
            record.Success = success;
            record.OutcomeTimestamp = DateTime.UtcNow;

            _logger.LogInformation("Recorded outcome for decision: {Decision}, Success: {Success}", decision, success);
        }
        else
        {
            _logger.LogWarning("No matching decision found for outcome: {Decision}", decision);
        }
    }

    /// <summary>
    /// Gets decision history
    /// </summary>
    /// <param name="count">The number of decisions to return</param>
    /// <returns>The decision history</returns>
    public List<DecisionRecord> GetDecisionHistory(int count)
    {
        return _decisionHistory
            .OrderByDescending(d => d.Timestamp)
            .Take(count)
            .ToList();
    }

    /// <summary>
    /// Gets decision success rate
    /// </summary>
    /// <returns>The success rate (0.0 to 1.0)</returns>
    public double GetDecisionSuccessRate()
    {
        var completedDecisions = _decisionHistory.Where(d => d.Outcome != null).ToList();

        if (completedDecisions.Count == 0)
        {
            return 0.0;
        }

        return (double)completedDecisions.Count(d => d.Success) / completedDecisions.Count;
    }

    /// <summary>
    /// Gets decision success rate by intuition type
    /// </summary>
    /// <returns>The success rates by intuition type</returns>
    public Dictionary<IntuitionType, double> GetDecisionSuccessRateByType()
    {
        var result = new Dictionary<IntuitionType, double>();

        foreach (var type in Enum.GetValues<IntuitionType>())
        {
            var completedDecisions = _decisionHistory
                .Where(d => d.IntuitionType == type && d.Outcome != null)
                .ToList();

            if (completedDecisions.Count > 0)
            {
                result[type] = (double)completedDecisions.Count(d => d.Success) / completedDecisions.Count;
            }
            else
            {
                result[type] = 0.0;
            }
        }

        return result;
    }

    /// <summary>
    /// Generates an intuition for a situation
    /// </summary>
    /// <param name="situation">The situation description</param>
    /// <param name="domain">The domain (optional)</param>
    /// <returns>The generated intuition</returns>
    public Intuition GenerateIntuition(string situation, string? domain = null)
    {
        try
        {
            _logger.LogInformation("Generating intuition for situation: {Situation}", situation);

            // Choose intuition type
            var intuitionType = ChooseIntuitionType(situation, domain);

            // Generate intuition based on type
            Intuition intuition;

            switch (intuitionType)
            {
                case IntuitionType.PatternRecognition:
                    intuition = _patternRecognition.GeneratePatternIntuition(situation, domain);
                    break;

                case IntuitionType.HeuristicReasoning:
                    intuition = _heuristicReasoning.GenerateHeuristicIntuition(situation, domain);
                    break;

                case IntuitionType.GutFeeling:
                    intuition = _gutFeeling.GenerateGutIntuition(situation);
                    break;

                default:
                    // Default to gut feeling
                    intuition = _gutFeeling.GenerateGutIntuition(situation);
                    break;
            }

            // Apply intuitive decision level to confidence
            intuition.Confidence *= _intuitiveDecisionLevel;

            _logger.LogInformation("Generated intuition: {Description} (Confidence: {Confidence:F2})",
                intuition.Description, intuition.Confidence);

            return intuition;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating intuition");

            // Return basic intuition
            return new Intuition
            {
                Id = Guid.NewGuid().ToString(),
                Description = "I have an intuition about this situation but can't articulate it clearly",
                Type = IntuitionType.GutFeeling,
                Confidence = 0.3 * _intuitiveDecisionLevel,
                Timestamp = DateTime.UtcNow,
                Source = "IntuitiveDecisionMaking"
            };
        }
    }
}

/// <summary>
/// Represents an intuitive decision
/// </summary>
public class IntuitiveDecision
{
    /// <summary>
    /// Gets or sets the decision description
    /// </summary>
    public string Decision { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the selected option
    /// </summary>
    public string SelectedOption { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the options
    /// </summary>
    public List<string> Options { get; set; } = new();

    /// <summary>
    /// Gets or sets the confidence (0.0 to 1.0)
    /// </summary>
    public double Confidence { get; set; } = 0.0;

    /// <summary>
    /// Gets or sets the intuition type
    /// </summary>
    public IntuitionType IntuitionType { get; set; }

    /// <summary>
    /// Gets or sets the timestamp
    /// </summary>
    public DateTime Timestamp { get; set; }

    /// <summary>
    /// Gets or sets the explanation
    /// </summary>
    public string Explanation { get; set; } = string.Empty;
}

/// <summary>
/// Represents a decision record
/// </summary>
public class DecisionRecord
{
    /// <summary>
    /// Gets or sets the decision description
    /// </summary>
    public string Decision { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the selected option
    /// </summary>
    public string SelectedOption { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the options
    /// </summary>
    public List<string> Options { get; set; } = new();

    /// <summary>
    /// Gets or sets the confidence (0.0 to 1.0)
    /// </summary>
    public double Confidence { get; set; } = 0.0;

    /// <summary>
    /// Gets or sets the intuition type
    /// </summary>
    public IntuitionType IntuitionType { get; set; }

    /// <summary>
    /// Gets or sets the timestamp
    /// </summary>
    public DateTime Timestamp { get; set; }

    /// <summary>
    /// Gets or sets the outcome
    /// </summary>
    public string? Outcome { get; set; }

    /// <summary>
    /// Gets or sets whether the decision was successful
    /// </summary>
    public bool Success { get; set; }

    /// <summary>
    /// Gets or sets the outcome timestamp
    /// </summary>
    public DateTime? OutcomeTimestamp { get; set; }
}

File: TarsEngine.FSharp.Core/Consciousness/Intelligence/Services/IntuitiveReasoning/IntuitiveDecisionMaking.fs
namespace TarsEngine.FSharp.Core.Consciousness.Intelligence.Services.IntuitiveReasoning

open System
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open TarsEngine.FSharp.Core.Consciousness.Intelligence

/// <summary>
/// Implementation of intuitive decision making methods.
/// </summary>
module IntuitiveDecisionMaking =
    /// <summary>
    /// Evaluates options using pattern recognition.
    /// </summary>
    /// <param name="options">The options.</param>
    /// <param name="patternRecognitionLevel">The pattern recognition level.</param>
    /// <param name="random">The random number generator.</param>
    /// <returns>The option scores.</returns>
    let evaluateOptionsWithPatternRecognition (options: string list) (patternRecognitionLevel: float) (random: Random) =
        options
        |> List.map (fun option ->
            // Base score is random but influenced by pattern recognition level
            let baseScore = 0.3 + (0.4 * random.NextDouble())
            
            // Pattern recognition bonus
            let patternBonus = 
                // Longer options might contain more recognizable patterns
                let lengthFactor = Math.Min(1.0, option.Length / 20.0) * 0.1
                
                // Options with certain keywords might trigger pattern recognition
                let keywordBonus =
                    let keywords = ["pattern"; "similar"; "like"; "recognize"; "familiar"; "seen"; "before"; "known"]
                    let optionLower = option.ToLowerInvariant()
                    keywords
                    |> List.sumBy (fun keyword -> if optionLower.Contains(keyword) then 0.05 else 0.0)
                
                // Combine factors
                (lengthFactor + keywordBonus) * patternRecognitionLevel
            
            // Calculate final score
            let finalScore = Math.Min(1.0, baseScore + patternBonus)
            
            (option, finalScore))
        |> Map.ofList
    
    /// <summary>
    /// Evaluates options using heuristic reasoning.
    /// </summary>
    /// <param name="options">The options.</param>
    /// <param name="heuristicReasoningLevel">The heuristic reasoning level.</param>
    /// <param name="random">The random number generator.</param>
    /// <returns>The option scores.</returns>
    let evaluateOptionsWithHeuristicReasoning (options: string list) (heuristicReasoningLevel: float) (random: Random) =
        options
        |> List.map (fun option ->
            // Base score is random but influenced by heuristic reasoning level
            let baseScore = 0.3 + (0.3 * random.NextDouble())
            
            // Apply various heuristics
            
            // Simplicity heuristic - shorter options are preferred
            let simplicityBonus = (10.0 - Math.Min(10.0, float option.Length / 5.0)) * 0.01
            
            // Familiarity heuristic - options with familiar terms are preferred
            let familiarityBonus =
                let optionLower = option.ToLowerInvariant()
                if optionLower.Contains("familiar") || 
                   optionLower.Contains("known") || 
                   optionLower.Contains("proven") then
                    0.1
                else
                    0.0
            
            // Availability heuristic - options that are easy to think of examples for
            let availabilityBonus =
                let optionLower = option.ToLowerInvariant()
                if optionLower.Contains("example") || 
                   optionLower.Contains("instance") || 
                   optionLower.Contains("case") then
                    0.1
                else
                    0.0
            
            // Apply heuristic reasoning level to bonuses
            let heuristicBonus = (simplicityBonus + familiarityBonus + availabilityBonus) * heuristicReasoningLevel
            
            // Calculate final score
            let finalScore = Math.Min(1.0, baseScore + heuristicBonus)
            
            (option, finalScore))
        |> Map.ofList
    
    /// <summary>
    /// Evaluates options using gut feeling.
    /// </summary>
    /// <param name="options">The options.</param>
    /// <param name="gutFeelingLevel">The gut feeling level.</param>
    /// <param name="random">The random number generator.</param>
    /// <returns>The option scores.</returns>
    let evaluateOptionsWithGutFeeling (options: string list) (gutFeelingLevel: float) (random: Random) =
        options
        |> List.map (fun option ->
            // Base score is mostly random for gut feeling
            let baseScore = 0.3 + (0.4 * random.NextDouble())
            
            // Gut feeling is more random but influenced by gut feeling level
            let gutBonus = (random.NextDouble() - 0.5) * gutFeelingLevel
            
            // Calculate final score
            let finalScore = Math.Min(1.0, Math.Max(0.0, baseScore + gutBonus))
            
            (option, finalScore))
        |> Map.ofList
    
    /// <summary>
    /// Evaluates options intuitively.
    /// </summary>
    /// <param name="options">The options.</param>
    /// <param name="intuitionType">The intuition type.</param>
    /// <param name="patternRecognitionLevel">The pattern recognition level.</param>
    /// <param name="heuristicReasoningLevel">The heuristic reasoning level.</param>
    /// <param name="gutFeelingLevel">The gut feeling level.</param>
    /// <param name="random">The random number generator.</param>
    /// <returns>The option scores.</returns>
    let evaluateOptionsIntuitively (options: string list) (intuitionType: IntuitionType option) 
                                  (patternRecognitionLevel: float) (heuristicReasoningLevel: float) 
                                  (gutFeelingLevel: float) (random: Random) =
        // Determine which intuition type to use
        let actualType =
            match intuitionType with
            | Some t -> t
            | None ->
                // Choose based on levels
                IntuitionGeneration.chooseIntuitionType patternRecognitionLevel heuristicReasoningLevel gutFeelingLevel random
        
        // Evaluate based on intuition type
        match actualType with
        | IntuitionType.PatternRecognition ->
            evaluateOptionsWithPatternRecognition options patternRecognitionLevel random
        | IntuitionType.HeuristicReasoning ->
            evaluateOptionsWithHeuristicReasoning options heuristicReasoningLevel random
        | IntuitionType.GutFeeling ->
            evaluateOptionsWithGutFeeling options gutFeelingLevel random
        | _ ->
            // For custom or unknown types, use a weighted combination
            let patternScores = evaluateOptionsWithPatternRecognition options patternRecognitionLevel random
            let heuristicScores = evaluateOptionsWithHeuristicReasoning options heuristicReasoningLevel random
            let gutScores = evaluateOptionsWithGutFeeling options gutFeelingLevel random
            
            // Combine scores with weights based on levels
            let totalLevel = patternRecognitionLevel + heuristicReasoningLevel + gutFeelingLevel
            let patternWeight = patternRecognitionLevel / totalLevel
            let heuristicWeight = heuristicReasoningLevel / totalLevel
            let gutWeight = gutFeelingLevel / totalLevel
            
            options
            |> List.map (fun option ->
                let patternScore = Map.find option patternScores
                let heuristicScore = Map.find option heuristicScores
                let gutScore = Map.find option gutScores
                
                let combinedScore = 
                    (patternScore * patternWeight) + 
                    (heuristicScore * heuristicWeight) + 
                    (gutScore * gutWeight)
                
                (option, combinedScore))
            |> Map.ofList
    
    /// <summary>
    /// Makes an intuitive decision.
    /// </summary>
    /// <param name="options">The options.</param>
    /// <param name="intuitionType">The intuition type.</param>
    /// <param name="patternRecognitionLevel">The pattern recognition level.</param>
    /// <param name="heuristicReasoningLevel">The heuristic reasoning level.</param>
    /// <param name="gutFeelingLevel">The gut feeling level.</param>
    /// <param name="random">The random number generator.</param>
    /// <returns>The selected option and the intuition.</returns>
    let makeIntuitiveDecision (options: string list) (intuitionType: IntuitionType option) 
                             (patternRecognitionLevel: float) (heuristicReasoningLevel: float) 
                             (gutFeelingLevel: float) (random: Random) =
        // Evaluate options
        let scores = evaluateOptionsIntuitively options intuitionType patternRecognitionLevel heuristicReasoningLevel gutFeelingLevel random
        
        // Determine which intuition type was used
        let actualType =
            match intuitionType with
            | Some t -> t
            | None ->
                // Choose based on levels
                IntuitionGeneration.chooseIntuitionType patternRecognitionLevel heuristicReasoningLevel gutFeelingLevel random
        
        // Find the option with the highest score
        let selectedOption =
            scores
            |> Map.toList
            |> List.maxBy snd
            |> fst
        
        // Get the score of the selected option
        let selectedScore = Map.find selectedOption scores
        
        // Generate an explanation based on the intuition type
        let explanation =
            match actualType with
            | IntuitionType.PatternRecognition ->
                sprintf "This option matches patterns I've observed in successful outcomes. It has a familiar structure that suggests it will work well."
            | IntuitionType.HeuristicReasoning ->
                sprintf "This option aligns with proven heuristics for this type of decision. It has qualities that are typically associated with good outcomes."
            | IntuitionType.GutFeeling ->
                sprintf "I have a strong intuitive sense that this is the right choice. While I can't fully articulate why, it feels like the best direction."
            | IntuitionType.Custom name ->
                sprintf "Using %s intuition, this option stands out as the most promising choice." name
            | _ ->
                sprintf "This option intuitively feels like the best choice based on a combination of factors."
        
        // Create an intuition for the decision
        let intuition = {
            Id = Guid.NewGuid().ToString()
            Description = sprintf "Intuitively, '%s' seems like the best option." selectedOption
            Type = actualType
            Confidence = selectedScore
            Timestamp = DateTime.UtcNow
            Context = Map.ofList [
                "Options", box options
                "Scores", box (scores |> Map.toList)
            ]
            Tags = ["decision"; "intuitive"; actualType.ToString().ToLowerInvariant()]
            Source = "Intuitive Decision Making"
            VerificationStatus = VerificationStatus.Unverified
            VerificationTimestamp = None
            VerificationNotes = ""
            Accuracy = None
            Impact = 0.7 // Decisions typically have high impact
            Explanation = explanation
            Decision = "Option Selection"
            SelectedOption = selectedOption
            Options = options
        }
        
        (selectedOption, intuition)
